src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/GCRefMapNode.cs

```diff
@@ -50,20 +50,19 @@ public void AppendMangledName(NameMangler nameMangler, Utf8StringBuilder sb)
 
         public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
         {
-            if (_methods.Count == 0 || relocsOnly)
-            {
-                return new ObjectData(
-                    data: Array.Empty<byte>(), 
-                    relocs: Array.Empty<Relocation>(), 
-                    alignment: 1, 
-                    definedSymbols: new ISymbolDefinitionNode[] { this });
-            }
-
             _methods.MergeSort(CompilerComparer.Instance);
             GCRefMapBuilder builder = new GCRefMapBuilder(factory.Target, relocsOnly);
+
             builder.Builder.RequireInitialAlignment(4);
             builder.Builder.AddSymbol(this);
 
+            if (_methods.Count == 0 || relocsOnly)
+            {
+                // Size header (R2RDump needs this)
+                builder.Builder.EmitInt(0);
```

Comment:
Instead of writing out dummy padding, can we write AuxiliaryData RVA as zero when the GCRefMap is empty?

Reply:
That should work. I'll try it, thanks!

Reply:
There are other cases where we attempt to serialize an empty gcrefmap, not just in the header. I added a Debug.Assert to check whether there were any methods and it's being hit during compilation of the BCL. I need to figure out whether it's correct to omit the 0 size header in that case or not.

Reply:
I tested this and with a `Debug.Assert(_methods.Count > 0);` in the GetObjectData, the assert still triggers when building the test suite even if I do the 'auxiliarydata RVA is zero' thing. So I think we need this path for the zero method count to write the size header.

Thank you for calling my attention to this code though because I think the size header is incorrect in relocsOnly mode. I've fixed that.

Reply:
Yes, GetData on gcRefMap will be called since GCRefMap is added as unconditional dependency in GetStaticDependencies. It means that we will try to include its content in the final image.

It should be fine for GetData to contribute no content like it was before this change. Can all changes in GetData be reverted (keep the changes to emit zero AuxiliaryData RVA)? Is there something that does not work when you do that?

BTW: There is also inefficiency one level up. We emit empty `READYTORUN_IMPORT_SECTION` entries that is unnecessary. If we stopped emitting empty `READYTORUN_IMPORT_SECTION` entries, the bug with bogus AuxiliaryData RVA would be fixed as well.

Reply:
It seems wholly wrong to ever emit content for a refmap that is missing the size header. If the size header isn't mandatory, why is it there? How can a consumer reading the refmap know that the data following the refmap isn't a refmap size header at runtime? Is there an implicit size header somewhere else that protects us, and R2RDump isn't reading it?